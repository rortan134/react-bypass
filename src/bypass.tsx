import { Slot, Slottable } from "@radix-ui/react-slot";
import * as React from "react";

type BypassElement = React.ElementRef<typeof Slot>;
interface BypassProps extends React.ComponentPropsWithoutRef<typeof Slot> {
    disabled?: boolean;
}

/**
 * Bypass component that skips one level of nesting in the component tree.
 *
 * @component
 * @example
 * // Basic usage
 * <Bypass>
 *   <div>
 *     <span>Content</span>
 *   </div>
 * </Bypass>
 * // Renders: <span>Content</span>
 *
 * @example
 * // With disabled prop
 * <Bypass disabled>
 *   <div>
 *     <span>Content</span>
 *   </div>
 * </Bypass>
 * // Renders: <div><span>Content</span></div>
 *
 * @param {Object} props - The component props
 * @param {boolean} [props.disabled=false] - When true, the component renders its children without bypassing
 * @param {React.ReactNode} props.children - The child elements to render
 *
 * @returns {React.ReactElement} The rendered component
 *
 * @throws {Error} Throws an error if multiple children are provided
 *
 * @see {@link https://www.radix-ui.com/primitives/docs/utilities/slot Radix UI Slot}
 */
const Bypass = React.forwardRef<BypassElement, BypassProps>((props, forwardedRef) => {
    const { disabled = false, children, ...rest } = props;
    if (disabled) {
        return children;
    }

    const nextNestedChildren = React.isValidElement(children)
        ? // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
          (React.Children.only(children).props.children as React.ReactNode)
        : null;

    const RenderNextNestedChildren = () => {
        return React.isValidElement(nextNestedChildren) ? (
            // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
            <Slottable>{nextNestedChildren}</Slottable>
        ) : typeof nextNestedChildren === "string" ? (
            nextNestedChildren
        ) : null;
    };

    return (
        <Slot {...rest} ref={forwardedRef}>
            <RenderNextNestedChildren />
        </Slot>
    );
});
Bypass.displayName = "Bypass";

export { Bypass };
export type { BypassElement, BypassProps };
