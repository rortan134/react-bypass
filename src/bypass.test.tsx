/**
 * @jest-environment jsdom
 */
import { fireEvent, render, screen } from "@testing-library/react";
import * as React from "react";
import { beforeEach, describe, expect, it, vi } from "vitest";
import { Bypass } from "./bypass.js";

describe("renders correctly", () => {
    it("should render the next nested children", () => {
        const tree = render(
            <Bypass>
                <div>
                    <span>foo</span>
                </div>
            </Bypass>
        );
        expect(tree.container.textContent).toBe("foo");
        expect(tree.container.firstElementChild).toMatchSnapshot();
    });
    it("should render a single button", () => {
        const tree = render(
            <Bypass>
                <button>
                    <button>Click me</button>
                </button>
            </Bypass>
        );
        expect(tree.container.textContent).toBe("Click me");
        expect(tree.container.firstElementChild).toMatchSnapshot();
    });
    it("should render the next nested children with multiple elements", () => {
        const tree = render(
            <Bypass>
                <div>
                    <section>
                        <div>
                            <span>bar</span>
                        </div>
                    </section>
                </div>
            </Bypass>
        );
        expect(tree.container.textContent).toBe("bar");
        expect(tree.container.firstElementChild).toMatchSnapshot();
    });
    it("should render if next nested children is text node", () => {
        const tree = render(
            <Bypass>
                <div>bar</div>
            </Bypass>
        );
        expect(tree.container.textContent).toBe("bar");
        expect(tree.container.firstElementChild).toBeNull();
    });
    it("should not render anything if there are no next nested children", () => {
        const tree = render(
            <Bypass>
                <div></div>
            </Bypass>
        );
        expect(tree.container.firstElementChild).toBeNull();
    });
});

describe("with interaction", () => {
    const handleClick = vi.fn();
    const handleChildClick = vi.fn();

    beforeEach(() => {
        handleClick.mockReset();
        handleChildClick.mockReset();
        render(
            <button type="button" onClick={handleClick}>
                <Bypass>
                    <button type="button" onClick={handleChildClick}>
                        <span>Click me</span>
                    </button>
                </Bypass>
            </button>
        );
        screen.getAllByRole("button").forEach((e) => fireEvent.click(e));
    });

    it("works with onClick", async () => {
        expect(handleClick).toHaveBeenCalledTimes(1);
        expect(handleChildClick).toHaveBeenCalledTimes(0);
    });
});
