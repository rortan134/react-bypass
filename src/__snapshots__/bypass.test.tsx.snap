// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders correctly > should render a single button 1`] = `
<button>
  Click me
</button>
`;

exports[`renders correctly > should render the next nested children 1`] = `
<span>
  foo
</span>
`;

exports[`renders correctly > should render the next nested children with multiple elements 1`] = `
<section>
  <div>
    <span>
      bar
    </span>
  </div>
</section>
`;
